content/2017/04/01/async-notifications.rst

```diff
@@ -0,0 +1,162 @@
+state it's in.
+
+The most simple (or simplistic) approach can be to use the database to store
+state. You can do this with Celery's database result backend, or a custom task
+state model that you periodically update. Imagine something like this:
+
+.. code:: python
+
+    @app.task
+    def my_task(some_arg):
+        # some unique identifier, that you can recover outside the task:
+        task_id = get_task_id_based_on_arg(some_arg)
+        state, _ = TaskState.objects.get_or_create(task_id=task_id)
+
+        total = len(some_arg)
+        for i, elem in enumerate(some_arg):
+            process_elem(elem)
+            # Every 100 elements, update the percentage processed:
+            if i % 100 == 0:
+                state.percent_done = (i / float(total)) * 100
+                state.save()
+
+Then in your views, you can retrieve the appropriate TaskState and show how
+much has been processed. Sometimes that's a good approach, but usually I think
+that's pretty clunky. It can thrash the database, leave records lying around if
+things die halfway through, and still doesn't give you a smooth experience;
+your user has to refresh to see updates.
+
+As an aside, it might be tempting to do something like this with the Django
+`messages framework`_. However, adding and retrieving messages requires the
+``request`` object. Even the pickle serializer can fail to serialize the
+request object. I would strongly recommend saving yourself the time and
+trouble, and using anything but messages for this.
+
+.. _messages framework: https://docs.djangoproject.com/en/1.10/ref/contrib/messages/
+
+So what if you want real-time updates? What if that page refresh is bumming you
+out?
+
+A nice option is something like `Pusher`_. They provide a service that you can
+push to from inside your app (in the request/response cycle, or in a background
+task) using a nice Python library, and some JS to get your users talking to
+their realtime websocket-y servers, to get those updates. Their JS library even
+includes sensible fallbacks for when websockets aren't available. The one
+caution is that their prices take a curve that can be a bit steep for some
+situations; if developer time is cheaper than ongoing service costs, then it
+might be worth rolling your own websocket solution. Which brings us to our next
+section.
+
+.. _Pusher: https://pusher.com/
+
+Django-Channels
+---------------
+
+If you are using the newer Django Channels package for background tasks, this
+has the added benefit of making it possible for you to make and manage your own
+websockets connections.
+
+For an overview of Channels itself, see the `channels docs`_ or Jacob
+Kaplan-Moss's `excellent blog post`_ on the subject.
+
+.. _channels docs: https://channels.readthedocs.io/en/stable/
+.. _excellent blog post: https://blog.heroku.com/in_deep_with_django_channels_the_future_of_real_time_apps_in_django
+
+The one thing you should know right here and now is that the
```

Comment:
I found this transition a bit jarring. You're off-loading the actual introduction to Django Channels (for those who aren't familiar with it already), and then jumping right into what initially sounds like technical details about a ``channel_routing`` attribute. Maybe at least provide a brief overview, or give me some sense of what ``channel_routing`` is and why I should care?